presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java

```diff
@@ -143,16 +162,41 @@ public Iterable<HivePartition> getPartitionsIterator(
         else {
             return () -> {
                 List<String> partitionNames = partitionFilteringFromMetastoreEnabled ? getFilteredPartitionNames(session, metastore, hiveTableHandle, effectivePredicate) : getAllPartitionNames(session, metastore, hiveTableHandle, constraint);
-                return partitionNames.stream()
-                        // Apply extra filters which could not be done by getFilteredPartitionNames
-                        .map(partitionName -> parseValuesAndFilterPartition(tableName, partitionName, partitionColumns, partitionTypes, constraint))
-                        .filter(Optional::isPresent)
-                        .map(Optional::get)
-                        .iterator();
+
+                if (isParallelPartitionValueParsingEnabled(session) && partitionNames.size() > PARTITION_NAMES_BATCH_SIZE) {
```

Comment:
If its not too complex, can we somehow unify these two branches to avoid any bugs. Maybe by playing games with the type of executor or such.

Reply:
The actual logic of how we extract the partition value is still the same code. The only thing that changed is whether we do it parallely or not. So should be fine i guess.